src/Everscale/Cell.cpp

```diff
@@ -0,0 +1,387 @@
+            const auto child = doneCells.find(raw.references[r]);
+            if (child == doneCells.end()) {
+                throw std::runtime_error("child cell not found");
+            }
+            references[r] = child->second;
+        }
+
+        auto cell = std::make_shared<Cell>(raw.bitLen, std::move(raw.data), raw.references.size(), std::move(references));
+        cell->finalize();
+        doneCells.emplace(index - 1, cell);
+    }
+
+    const auto root = doneCells.find(rootIndex);
+    if (root == doneCells.end()) {
+        throw std::runtime_error("root cell not found");
+    }
+    return std::move(root->second);
+}
+
+class SerializationContext {
+public:
+    static SerializationContext build(const Cell& cell) {
+        SerializationContext ctx{};
+        fillContext(cell, ctx);
+        return ctx;
+    }
+
+    void encode(Data& os) const {
+        os.reserve(os.size() + serializedSize);
+
+        const auto cellCount = static_cast<uint16_t>(reversedCells.size());
+
+        // Write header
+        encode32BE(BOC_MAGIC, os);
+        os.push_back(REF_SIZE);
+        os.push_back(OFFSET_SIZE);
+        encode16BE(static_cast<uint16_t>(cellCount), os);
+        encode16BE(1, os); // root count
+        encode16BE(0, os); // absent cell count
+        encode16BE(static_cast<uint16_t>(serializedSize), os);
+        encode16BE(0, os); // root cell index
+
+        // Write cells
+        size_t i = cellCount - 1;
+        while (true) {
+            const auto& cell = *reversedCells[i];
+
+            // Write cell data
+            const auto [d1, d2] = cell.getDescriptorBytes();
+            os.push_back(d1);
+            os.push_back(d2);
+            os.insert(os.end(), cell.data.begin(), cell.data.end());
+
+            // Write cell references
+            for (const auto& child : cell.references) {
+                if (child == nullptr) {
+                    break;
+                }
+
+                // Map cell hash to index (which must be presented)
+                const auto it = indices.find(child->hash);
+                assert(it != indices.end());
+
+                encode16BE(cellCount - it->second - 1, os);
+            }
+
+            if (i == 0) {
+                break;
+            } else {
+                --i;
+            }
+        }
+    }
+
+private:
+    // uint16_t will be enough for wallet transactions (e.g. 64k is the size of the whole elector)
+    constexpr static uint8_t REF_SIZE = 2;
+    constexpr static uint8_t OFFSET_SIZE = 2;
+
+    // NOTE: Initialized with header size
+    size_t serializedSize =
+        /*magic*/ sizeof(BOC_MAGIC) +
+        /*ref_size*/ 1 +
+        /*offset_size*/ 1 +
+        /*cell_count*/ REF_SIZE +
+        /*root_count*/ REF_SIZE +
+        /*absent_count*/ REF_SIZE +
+        /*data_size*/ OFFSET_SIZE +
+        /*root_cell_index*/ REF_SIZE;
+
+    uint16_t index = 0;
+    std::map<Cell::CellHash, uint16_t> indices{};
```

Comment:
same question

Reply:
We use array as key that's why we preferred map to unordered_map